src/hardware/serialport/serialport.cpp

```diff
@@ -1232,38 +1238,31 @@ bool CSerial::Getchar(uint8_t *data, uint8_t *lsr, bool wait_dsr, uint32_t timeo
 bool CSerial::Putchar(uint8_t data, bool wait_dsr, bool wait_cts, uint32_t timeout)
 {
 	const double starttime = PIC_FullIndex();
-	// wait for it to become empty
-	while (!(Read_LSR() & LSR_TX_HOLDING_EMPTY_MASK)) {
-		CALLBACK_Idle();
+	bool timed_out = false;
+	auto status = Read_MSR();
+
+	// First wait until our transmit queue is empty
+	while (!timed_out && !(status & LSR_TX_HOLDING_EMPTY_MASK)) {
+		timed_out = idle(starttime, timeout);
+		status = Read_MSR();
 	}
-	// wait for DSR+CTS on
-	if (wait_dsr || wait_cts) {
-		if (wait_dsr && wait_cts) {
-			constexpr auto dsr_cts_mask = MSR_DSR_MASK | MSR_CTS_MASK;
-			while (((Read_MSR() & dsr_cts_mask) != dsr_cts_mask) &&
-			       (starttime > PIC_FullIndex() - timeout)) {
-				CALLBACK_Idle();
-			}
-		} else if (wait_dsr) {
-			while (!(Read_MSR() & MSR_DSR_MASK) &&
-			       (starttime > PIC_FullIndex() - timeout)) {
-				CALLBACK_Idle();
-			}
-		} else if (wait_cts) {
-			while (!(Read_MSR() & MSR_CTS_MASK) &&
-			       (starttime > PIC_FullIndex() - timeout)) {
-				CALLBACK_Idle();
-			}
-		}
-		if (!(starttime > PIC_FullIndex() - timeout)) {
+
+	// If waits are disabled, then the ready-checks are simply true.
+	// Otherwise we check the actual line status and keep waiting until
+	// their ready bits are present.
+	while (!timed_out && (!(wait_dsr ? status & MSR_DSR_MASK : true) ||
+	                      !(wait_cts ? status & MSR_CTS_MASK : true))) {
+		timed_out = idle(starttime, timeout);
+		status = Read_MSR();
+	}
```

Comment:
For the context: I tried to review this commit 2 times already, and I get stuck each time, because this is not a drop-in replacement nor "simplification". This commit refactors the code and changes the behaviour (mostly for better, but I am not 100% convinced yet).

Until I will be able to express my proper reservations or agree that this is definitely better code, here are some improvements needed:

1. Commit message needs to reflect that this is pretty serious refactor and we are changing the behaviour *on purpose*.
2. In these specific lines I highlighted here - I suggest replacing this complex check for `while` by introducing wait flags earlier, something like:

```C++
const uint32_t wait_flags = (wait_dsr ? MSR_DSR_MASK : 0x0) | (wait_cts ? MSR_CTS_MASK : 0x0);
```
3. I know this code is currently used only with both `wait_dsr` and `wait_cts` set to `true`, but the behaviour of code was changed for cases when both are set to `false` (which might introduce very, very hard-to-detect regression if upstream introduces some new code in here in the future).

Reply:
Thanks for going through this and recommending next steps. 

For 1), It is a serious refactoring, however functionally the behaviour is identical.  I will work on some code to show that's the case; and then we can make the commit message reflect that. 

For 2), This will make the while logic cleaner, but we'll need two places where we assign wait_flags.  I guess that's ok? (once before the while, then again inside the while).

Wait_flags could then be combined with the first LSR_TX_HOLDING_EMPTY_MASK check as well, further simplifying this function.  Simply put, all the criteria need to be achieved before we stop waiting.. and it's fine to check all of it at the same time.

Reply:
Re 2: No, why? I think a single assignment is enough (only for loop in 1253). Subsequent (and earlier) calls to `Read_MSR` should probably use `|=` instead of `=` though. But I might be missing something…

Reply:
Sorry @dreamer ; you're right. 

I had thought `wait_flags` included the assessment of Read_MSR (as one of the variables); but I didn't look at it carefully enough. 

This is indeed a very nice suggestion and will clean it up even more.  
(Working on the refactored comparison right now..  and adding this now :-)

Reply:
Test code: https://onlinegdb.com/YwanLSBbx

Results after testing 10000 random scenarios (repeatedly).
- ref1 missed: 0 (this is the existing refactored code in this PR)
- ref2 missed: ~3000 or so (this is using the `wait_flags`)

I would prefer to use the `wait_flags` though, as it's even more elegant - maybe it needs to be split into two and `or`'d?

``` c++
/***************************************************************
Compare refactored DSR and CTS wait logic versus original logic.
***************************************************************/
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string>

#define MSR_CTS_MASK 0x10
#define MSR_DSR_MASK 0x20

bool orig_wait(bool dsr, bool cts, unsigned int msr)
{
	if (dsr || cts) {
		if (dsr && cts) {
			constexpr auto dsr_cts_mask = MSR_DSR_MASK | MSR_CTS_MASK;
			if ((msr & dsr_cts_mask) != dsr_cts_mask) {
				return true;
			}
		} else if (dsr) {
			if (!(msr & MSR_DSR_MASK)) {
				return true;
			}
		} else if (cts) {
			if (!(msr & MSR_CTS_MASK)) {
				return true;
			}
		}
	}
	return false;
}

bool ref1_wait(bool dsr, bool cts, unsigned int msr)
{
	if (!(dsr ? msr & MSR_DSR_MASK : true) || !(cts ? msr & MSR_CTS_MASK : true))
		return true;
	return false;
}

bool ref2_wait(bool dsr, bool cts, unsigned int msr)
{
	const unsigned int flags = (dsr ? MSR_DSR_MASK : 0x0) |
	                           (cts ? MSR_CTS_MASK : 0x0);
	if (!(msr & flags))
		return true;
	return false;
}

std::string assess(const bool orig, const bool ref1, int &missed1, bool ref2, int &missed2)
{
	if (ref1 == orig && ref2 == orig) {
		return "both passed";
	} else if (ref1 == orig) {
		missed2++;
		return "ref1 passed";
	} else if (ref2 == orig) {
		missed1++;
		return "ref2 passed";
	}
	missed1++;
	missed2++;
	return "both failed";
}

int main()
{
	srand(888);
	int i = 0;
	std::string result;
	int missed1 = 0;
	int missed2 = 0;

	while (++i < 10000) {
		const bool dsr = rand() % 2;
		const bool cts = rand() % 2;
		const unsigned int msr = rand() % UINT_MAX;
		const bool orig = orig_wait(dsr, cts, msr);
		const bool ref1 = ref1_wait(dsr, cts, msr);
		const bool ref2 = ref2_wait(dsr, cts, msr);
		const std::string result = assess(orig, ref1, missed1, ref2, missed2);
		printf("[%s] dsr %d, cts %d, MSR=%-10u, orig %d, ref1 %d ref2 %d\n",
		       result.c_str(), dsr, cts, msr, orig, ref1, ref2);
	}
	printf("\nTested %d scenarios. ref1 missed: %d, ref2 missed: %d\n", i,
	       missed1, missed2);
	return 0;
}
```

Reply:
@dreamer , you also mentioned, _nor is it a "simplification"_, so I want to address that too. The code should be smaller and more readable. 

The bottom line understanding for this function is that we have three criteria, all of which need to be met otherwise we simply keep waiting. 
- our transfer queue is empty
- the receiver says it's 'ready' (dsr high), provided dsr is enabled on the line
- the receiver says we are 'clear to send' (cts high), provided cts is enabled on the line. 

There's probably a very nice way we can express this and make the code very understandable.

Reply:
Will push an update shortly; hopefully addresses most of noted points.

Reply:
Test code is identical to original, with the PRs refactoring (before), as well as with your suggested flags :-)
https://onlinegdb.com/yesKi3YoC